src/modules/fancyzones/lib/FancyZones.cpp

```diff
@@ -515,49 +524,114 @@ LRESULT CALLBACK FancyZones::s_WndProc(HWND window, UINT message, WPARAM wparam,
+        auto & newDevice = capturedScope->devices.emplace_back(device{L"FallbackDevice", monitor});
 
-                auto strongThis = reinterpret_cast<FancyZones*>(data);
-                strongThis->AddZoneWindow(monitor, deviceId);
-            }
+        std::array<wchar_t, 256> parsedId{};
+        if (!ParseDeviceId(displayDevice.DeviceID, parsedId.data(), size(parsedId)))
+        {
+            return TRUE;
+        }
+        // Get a generated part of the id, e.g. "5&25664547&0&UID4355" from "VSCBD34#5&25664547&0&UID4355"
+        std::wstring_view parsedIdView{parsedId.data(), wcslen(parsedId.data())};
+        newDevice._deviceID = parsedIdView;
+        if (parsedIdView.find(L'#') != std::wstring::npos)
+        {
+            // Check if EnumDisplayMonitors returned a duplicate data, it's an API bug and we should workaround it
+            const auto [_, hasUniqueID] = capturedScope->discoveredDeviceIDs.emplace(parsedIdView);
+            capturedScope->duplicateDeviceIdFound = !hasUniqueID || capturedScope->duplicateDeviceIdFound;
         }
         return TRUE;
     };
+    EnumDisplayMonitors(nullptr, nullptr, enumDisplayMonitorsCallback, reinterpret_cast<LPARAM>(&localScope));
+    std::vector<WmiMonitorID> wmiMonitorInfo;
+    try
+    {
+        m_wmiConnection.select_all(L"WmiMonitorID", [&](std::wstring_view xml_obj) {
+          wmiMonitorInfo.emplace_back(parse_monitorID_from_dtd(xml_obj));
+        });
+    }
+    catch(...) {}
 
+    if (GetSystemMetrics(SM_REMOTESESSION))
     {
-        std::unique_lock writeLock(m_lock);
-        m_zoneWindowMap.clear();
+        // When running in a remote session, devices have "Default_Monitor#<transient_part>" DeviceIDS.
+        // Since a remote session can have multiple devices, we can't just assign them a single "REMOTE" ID.
+        // We'll assign them unique numeric IDs instead.
+        for (size_t i = 0; i < size(localScope.devices); ++i)
+        {
+            wchar_t remoteDeviceID[32];
+            swprintf_s(remoteDeviceID, L"REMOTE%zu", i);
+            AddZoneWindow(localScope.devices[i]._handle, remoteDeviceID);
+        }
+        return;
+    }
+    
+    // Check if there's WMI data, if no, we're most likely running inside a VM
+    const bool hasWmiData = size(localScope.devices) == size(wmiMonitorInfo);
+    if (!hasWmiData)
+    {
+        // TODO: add more robust VM detection logic
+        for (const auto & dev : localScope.devices)
+        {
+            AddZoneWindow(dev._handle, dev._deviceID.c_str());
+        }
+        return;
+    }
+    // If EnumDisplayMonitors returned duplicate results, use WMI data
+    if (localScope.duplicateDeviceIdFound)
+    {
+        for (size_t i = 0; i < size(localScope.devices); ++i)
+        {
+            AddZoneWindow(localScope.devices[i]._handle, wmiMonitorInfo[i].hardware_id().c_str());
+        }
+        return;
+    }
+    for (const auto & dev : localScope.devices)

```

Comment:
Add a comment, something like:
// Default case: a bare metal machine and EnumDisplayMonitors() returned valid data